scripts/dev/upstream_merge/upstream_merge.sh

```diff
@@ -0,0 +1,206 @@
+#!/bin/bash
+
+# This script can be used for merging latest upstream for each of the
+# repos described in repos.conf
+#
+# For each local repo, it does following
+# 1. Check if local_base_branch specified in repos.conf exists
+# 2. Checkout local_base_branch and run git pull
+# 3. Add a 'remote' named 'automerge_upstream'
+#    a. If the remote already exists, remove it first
+# 4. Pulls remote branch into a branch named 'dev/automerge/upstream/<upstream_branch>'
+#    a. If the branch name already exists, remove it first
+# 5. Creates a branch named 'dev/automerge/ni' from 'nilrt/master/hardknott' and checkout
+#    a. If the branch name exists, ask to delete or skip or cancel
+# 6. Merges 'dev/automerge/upstream' into 'dev/automerge/ni'
+#    a. Success or failure is reported; script continues in either case
+
+DIR=$(dirname ${BASH_SOURCE})
+CONF_FILE="$DIR/repos.conf"
+
+REMOTE_REPO_NAME="automerge_upstream"
+LOCAL_UPSTREAM_BRANCH_PREFIX="dev/automerge/upstream"
+LOCAL_BRANCH_NAME="dev/automerge/ni"
+
+usage() {
+   echo "Usage: $0 [-c <conf file>][-h] " 1>&2
+   exit 1
+}
+
+parse_args() {
+   local OPTIND o
+   while getopts "c:h" o; do
+      case "${o}" in
+         c)
+            CONF_FILE=${OPTARG}
+            ;;
+         h)
+            usage
+            ;;
+         *)
+            usage
+            ;;
+      esac
+   done
+   shift $((OPTIND-1))
+}
+
+sanity_test_repo() {
+   LOCAL_BASE_BRANCH=$1
+   if ! `git rev-parse --verify $LOCAL_BASE_BRANCH &> /dev/null`; then
+      echo ""
+      echo "    Branch $LOCAL_BASE_BRANCH does not exist. Exiting"
+      exit 1
+   fi
+}
+
+update_local_base_branch() {
+   LOCAL_BASE_BRANCH=$1
+   if `git checkout $LOCAL_BASE_BRANCH &> /dev/null`; then
+      if ! `git pull &> /dev/null`; then
+         echo ""
+         echo "    Error pulling latest on $LOCAL_BASE_BRANCH. Exiting"
+         exit 1
+      fi
+   else
+      # Error changing branch
+      echo ""
+      echo "    Error switching to branch $LOCAL_BASE_BRANCH. Exiting"
+      exit 1
+   fi
+}
+
+add_remote() {
+   UPSTREAM_REPO=$1
+
+   git remote remove $REMOTE_REPO_NAME &> /dev/null || true
+   git remote add $REMOTE_REPO_NAME $UPSTREAM_REPO
+}
```

Comment:
I'm uneasy about this script modifying a submodule's remotes as a consequence of its action. I think we should honor however the user has their remotes configured, even if that means that the user needs to specify what they have named their upstream and NI GH remotes.

Reply:
My intent with this script was that however the user have configured their repos, the script would setup whatever it needs in its own namespace (`automerge` in this case. i.e., remotes are `automerge_upstream`, and branches are `dev/automerge/..`) and then do the merge.

On a newly setup nilrt repo, the user doesn't have upstream remotes setup and I don't think it should be a requirement for them to do that to use this script as that can be automated as well.

> I'm uneasy about this script modifying a submodule's remotes as a consequence of its action

The script only adds a new remote and shouldn't affect what the developer has configured for their use. The remote setup here is intended to be used only in the script (hence the namespace).

If your concern is about polluting remote list, we can consider adding a `--cleanup` option to the script that would go ahead and remove the remote (and `dev/automerge/...` branches too?).

Reply:
I would also like the setting up of remotes for all the OE layers to be automated, and not something I should have to do separately.